src/IO/RemoteReadBufferCache.cpp

```diff
@@ -0,0 +1,471 @@
+{
+    return !(file_metadata_ptr->getVersion() == file_metadata_->getVersion());
+}
+
+void RemoteCacheController::startBackgroundDownload(std::unique_ptr<ReadBuffer> in_readbuffer_, BackgroundSchedulePool & thread_pool)
+{
+    data_file_writer = std::make_unique<WriteBufferFromFile>((fs::path(local_path) / "data.bin").string());
+    flush(true);
+    ReadBufferPtr in_readbuffer(in_readbuffer_.release());
+    download_task_holder = thread_pool.createTask("download remote file",
+            [this, in_readbuffer]{ backgroundDownload(in_readbuffer); });
+    download_task_holder->activateAndSchedule();
+}
+
+void RemoteCacheController::backgroundDownload(ReadBufferPtr remote_read_buffer)
+{
+    file_status = DOWNLOADING;
+    size_t before_unflush_bytes = 0;
+    size_t total_bytes = 0;
+    while (!remote_read_buffer->eof())
+    {
+        size_t bytes = remote_read_buffer->available();
+
+        data_file_writer->write(remote_read_buffer->position(), bytes);
+        remote_read_buffer->position() += bytes;
+        total_bytes += bytes;
+        before_unflush_bytes += bytes;
+        if (before_unflush_bytes >= local_cache_bytes_read_before_flush)
+        {
+            std::unique_lock lock(mutex);
+            current_offset += total_bytes;
+            total_bytes = 0;
+            flush();
+            lock.unlock();
+            more_data_signal.notify_all();
+            before_unflush_bytes = 0;
+        }
+    }
+    std::unique_lock lock(mutex);
+    current_offset += total_bytes;
+    file_status = DOWNLOADED;
+    flush(true);
+    data_file_writer.reset();
+    lock.unlock();
+    more_data_signal.notify_all();
+    RemoteReadBufferCache::instance().updateTotalSize(file_metadata_ptr->getFileSize());
+    LOG_TRACE(log, "Finish download into local path: {}, file metadata:{} ", local_path.string(), file_metadata_ptr->toString());
+}
+
+void RemoteCacheController::flush(bool need_flush_status)
+{
+    if (data_file_writer)
+    {
+        data_file_writer->sync();
+    }
+    if (need_flush_status)
+    {
+        auto file_writer = std::make_unique<WriteBufferFromFile>(local_path / "info.txt");
+        Poco::JSON::Object jobj;
+        jobj.set("file_status", static_cast<Int32>(file_status));
+        jobj.set("metadata_class", metadata_class);
+        std::stringstream buf; // STYLE_CHECK_ALLOW_STD_STRING_STREAM
+        jobj.stringify(buf);
+        file_writer->write(buf.str().c_str(), buf.str().size());
+        file_writer->close();
+    }
+}
+
+RemoteCacheController::~RemoteCacheController()
+{
+    if (download_task_holder)
+        download_task_holder->deactivate();
+}
+
+void RemoteCacheController::close()
+{
+    // delete directory
+    LOG_TRACE(log, "Removing the local cache. local path: {}", local_path.string());
+    std::filesystem::remove_all(local_path);
+}
+
+std::unique_ptr<ReadBufferFromFileBase> RemoteCacheController::allocFile()
+{
+    ReadSettings settings;
+    settings.local_fs_prefetch = false;
```

Comment:
This is not needed. This setting is used only for reading of merge tree tables.

Reply:
`settings.local_fs_method` is pread in default, It seems would broke the rule of waiting enough data before reading in